### PR TITLE
feat(table): radio button on single select row

### LINKED ISSUE #3394: [Component Proposal] Radio buttons on data table
### What package is this for?

- [x] React
- [ ] Angular

### Summary
Radio button to enable single selection on data tables as currently there only exist multi selection capability with checkboxes. 
• Add radio button styling prop, mutually exclusive with the blue line styling that appears on the start of the row
• Consider allowing the radio button for data table is to allow a null state. So, on-load, the table has nothing selected, and I can click the currently selected radio to go back to null
Implementing this pattern in the design for Maximo Manager Role-Based Application - Maintenance manager for Release 1

**Additional context** 
This pattern currently exists in Carbon ([usage guidance ](https://www.carbondesignsystem.com/components/data-table/usage/#radio-selection)and [storybook example](https://react.carbondesignsystem.com/?path=/story/components-datatable-selection--with-radio-selection))

Invision design prototype for Maintenance manager utilizing this radio selection pattern: https://ibm.invisionapp.com/d/main#/console/15384087/320231588/preview

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release? Maximo Manager Role-Based Application - Maintenance manager for Release 1

_NB: The core contributors will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### You will consider this component complete when...

- [ ] It's available as an export of this library
- [ ] Additional criteria you may have [e.g. Fully responsive, Accessible to
      WCAG 2.0, etc.]

### Available extra resources

What resources do you have to assist this effort?

_Carbon add-ons for Watson IoT is a collaborative system. We encourage teams to
build components and submit them for integration._

### PR DESCRIPTION
**Pull request by jessieyan**
Closes #3394 

**Summary**

- single selection with radio button 
- there are 2 ways of single selection, it's either with a radio button or with a blue indicator.

**Change List (commits, features, bugs, etc)**

- add radio button when `hasRowSelection` is `single` and  option prop `showRadioButton` is `true`

**Acceptance Test (how to verify the PR)**

image

image

image

image

image

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Comment by JordanWSmith15** [comment]
@jessieyan I see an issue with the column headers having shifted over:

**Comment by jessieyan** [comment]
> @jessieyan I see an issue with the column headers having shifted over: 

Now it should be fixed.

Single select with radio button.
image

Single select without radio button.
image

**Comment by JordanWSmith15** [comment]
@jessieyan Can we get a deploy preview?

**Comment by jessieyan** [comment]
@JordanWSmith15 Hi Jordan, deploy preview is here https://deploy-preview-3423--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table--with-selection-and-batch-actions

File: packages/react/src/components/Table/Table.jsx [hunk 1]
```diff
@@ -991,6 +991,7 @@ const Table = (props) => {
                   'hasRowNesting',
                   'hasSingleRowEdit',
                   'hasRowSelection',
+                  'showRadioButton',
```
Comment (davidicus) [inline packages/react/src/components/Table/Table.jsx:994]: we may want to change the name of this prop to be more informative on where we are showing the radio button. `useRadioButtonSingleSelect` or something like that.

File: packages/react/src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx [hunk 2]
```diff
@@ -236,6 +245,27 @@ const TableBodyRow = ({
           />
         </span>
       </TableCell>
+    ) : hasRowSelection === 'single' && showRadioButton ? (
+      <TableCell
+        className={`${prefix}--radiobutton-table-cell`}
+        key={`${id}-row-selection-cell`}
+        onChange={isSelectable !== false ? () => onRowSelected(id, !isSelected) : null}
+        onClick={(e) => e.stopPropagation()}
```
Comment (davidicus) [inline packages/react/src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx:253]: In the story, the cell around the radio button does not trigger a row click or a row selection. Is this stopPropagation causing that?

**Review by davidicus — CHANGES_REQUESTED** [review]
Hi @jessieyan This PR looks pretty good, I just have two minor comments to address before we merge.

**Comment by jessieyan** [comment]
> Additionally can we look into being able to deselect the row in our StatefulTable?

Added deselect on the single selection rows. When selected and deselect, `isSelected` is false and `selectedIds` is empty.

File: packages/react/src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx (hunk 2 above)
Comment (jessieyan) [reply packages/react/src/components/Table/TableBody/TableBodyRow/TableBodyRow.jsx:253]: > In the story, the cell around the radio button does not trigger a row click or a row selection. Is this stopPropagation causing that?

This is the same design as the checkbox in `multi` select. It looks like we can select either the table row or the checkbox/radio button.  Do we want to update it?

**Review by davidicus — APPROVED** [review]

**Pull request merged**